Add UseStandardPixelAddressing option

Jjagg opened the pull request:
This implements an option to enable/disable half pixel offset as discussed in #6619. 

Currently this does not yet actually apply the half pixel offset for MG DX (behavior isn't changed). That's a bit more intricate to do because we have to modify 2MGFX to support the offset in user-written vertex shaders.

This does implement the option for OpenGL platforms. Default behavior did not change, but the `GraphicsDeviceManager.UseStandardPixelAddressing` flag can be set to `true` to render like the DX backend does (different from XNA).

Jjagg commented:
#6623 should get merged before this one.

Jjagg commented:
This is ready for review @cra0zy @tomspilman @roy-t @mrhelmut. 
This PR changes nothing for DX platforms and by default nothing on GL platforms. However users can make rendering on GL platforms consistent with DX by setting the `UseStandardPixelAddressing` flag to `true`.

Jjagg commented:
@cra0zy So what do you propose then as an alternative? We want people to set the flag to true.

cra0zy commented:
> @cra0zy So what do you propose then as an alternative? We want people to set the flag to true.

Why not set it to true by default and write a documentation page about it?

Apostolique commented:
Having it set to true as default would also be my choice. MonoGame aims to be consistent across platforms. This is a step in the right direction.

cra0zy commented:
In the meantime you can modify the templates back and write the doc page.

mrhelmut commented:
Since it doesn't break SpriteBatch and makes everything on par with DirectX, I'm totally fine with breaking the default value.

Jjagg commented:
Changes made, this is good to go if you ask me.

nkast commented on Test/Framework/Graphics/GraphicsDeviceTestFixtureBase.cs:
```diff
@@ -51,6 +51,11 @@ internal class GraphicsDeviceTestFixtureBase
         {
+#if !XNA
+            // We enable the half-pixel offset for XNA compatibility
+            GraphicsDeviceManager.UseStandardPixelAddressing = false;
```
if SpriteBatch rendering is not affected by the flag and produces the same results, then we don't need that. Shouldn't we instead test MG with the default setting?

Jjagg replied:
The tests that failed were not SpriteBatch tests. I don't really know why they did fail though.

 But we do want to set this flag in the tests for when we add tests for the built in effects (if we ever get around to that). Of course DX should support manipulating half pixel offset at runtime for it to work.

Jjagg commented:
@tomspilman Need your approval for this one because we changed the default to be non-XNA compatible (also added a doc page for XNA compatibility that explains the flag).

Apostolique commented:
I'd really like this to get merged. It makes MonoGame be more consistent between OpenGL and DirectX.

Jjagg commented:
For DX we should do the same as GL. That is, inject some code into shaders and set a flag/value in the shader that controls the offset based on the UseStandardPixelAddressing flag. 

Here's an interesting blog post that fixes DX9 shaders by injecting shader bytecode (the reverse of what we want to do, but the approach is the same): https://aras-p.info/blog/2016/04/08/solving-dx9-half-pixel-offset/.

Jjagg commented:
@cra0zy Could you merge this?

cra0zy commented:
Of course :)

Jjagg commented:
Thanks!

nkast commented:
I don't think that GraphicsDeviceManager is a good place for UseStandardPixelAddressing. It should go under GraphicsDevice. We should avoid a dependency from GraphicsDevice to the GraphicsDeviceManager.

GraphicsDeviceManager is just an implementation of IGraphicsDeviceManager/IGraphicsDeviceService, there are cases where you have a GraphicsDevice without a GraphicsDeviceManager/Game. e.g. in an editor or for GPU data processing.

Jjagg commented:
It doesn't really matter if you don't have a GraphicsDeviceManager because the property is static. But I'm fine with it in GraphicsDevice as well. It was put in GDM because that's where you set the configuration before creating the GraphicsDevice. If we do move if, we should definitely do it before a release.

nkast commented:
>> It was put in GDM because that's where you set the configuration before creating the GraphicsDevice. 

I see. A more complete implementation would be with a  'GraphicsDeviceManager.PrefferedUseStandardPixelAddressing', pass the value to the construction Parameters (which is how I would set it in my my Game-less editor), and set a fixed/readonly value for GraphicsDevice.UseStandardPixelAddressing for the entire life of GraphicsDevice. 

With this setup, we can have, 
 * Each instance of  GraphicsDevice has it's own setting.
 * The creator of GraphicsDevice is in control, 3rd party libraries can't change/force their preferred settings.
 * We can pre-calculate most of projection matrix values in SpriteEffect.

In the sort term thou , I can move it to GraphicsDevice and make it non-static.
